File: app/src/main/java/com/jakebarnby/filemanager/sources/local/LocalLoaderTask.java
package com.jakebarnby.filemanager.sources.local;

import android.os.Environment;

import com.jakebarnby.filemanager.sources.LoaderTask;
import com.jakebarnby.filemanager.sources.SourceListener;
import com.jakebarnby.filemanager.sources.models.Source;
import com.jakebarnby.filemanager.sources.models.SourceFile;
import com.jakebarnby.filemanager.util.TreeNode;
import com.jakebarnby.filemanager.util.Utils;

import java.io.File;

/**
 * Created by Jake on 8/2/2017.
 */

public class LocalLoaderTask extends LoaderTask {

    public LocalLoaderTask(Source source, SourceListener listener) {
        super(source, listener);
    }

    @Override
    protected Object initRootNode(String path) {
        File rootFile = new File(path);
        SourceFile rootSourceFile = new LocalFile(rootFile, mSource.getSourceName());
        mRootTreeNode = new TreeNode<>(rootSourceFile);
        mCurrentNode = mRootTreeNode;
        mSource.setCurrentDirectory(mRootTreeNode);
        mSource.setQuotaInfo(Utils.getStorageStats(new File(path)));
        return rootFile;
    }

    @Override
    protected TreeNode<SourceFile> readFileTree(Object rootObject) {
        if (rootObject != null && rootObject instanceof File) {
            File node = (File)rootObject;
            File listFile[] = node.listFiles();
            long dirSize = 0L;
            if (listFile != null && listFile.length > 0) {
                for (File file : listFile) {
                    SourceFile sourceFile = new LocalFile(file, mSource.getSourceName());
                    if (file.isDirectory()) {
                        mCurrentNode.addChild(sourceFile);
                        mCurrentNode = mCurrentNode.getChildren().get(mCurrentNode.getChildren().size() - 1);
                        readFileTree(file);
                        mCurrentNode.getParent().getData().addSize(mCurrentNode.getData().getSize());
                        mCurrentNode = mCurrentNode.getParent();
                    } else {
                        dirSize += file.length();
                        mCurrentNode.addChild(sourceFile);
                    }
                }
                mCurrentNode.getData().addSize(dirSize);
            }
        }
        return mRootTreeNode;
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/sources/local/LocalLoaderTask.kt
package com.jakebarnby.filemanager.sources.local

import com.jakebarnby.filemanager.models.Source
import com.jakebarnby.filemanager.models.SourceFile
import com.jakebarnby.filemanager.ui.sources.SourceFragmentContract
import com.jakebarnby.filemanager.util.TreeNode
import com.jakebarnby.filemanager.util.Utils
import com.jakebarnby.filemanager.workers.LoaderTask
import java.io.File

/**
 * Created by Jake on 8/2/2017.
 */
class LocalLoaderTask(
    onStart: () -> Unit,
    onComplete: (TreeNode<SourceFile>) -> Unit
) : LoaderTask(onStart, onComplete) {

    override fun initRootNode(path: String): Any? {
        val rootFile = File(path)
        val rootSourceFile: SourceFile = LocalFile(rootFile, source.sourceId)
        rootTreeNode = TreeNode(rootSourceFile)
        currentNode = rootTreeNode
        source.currentDirectory = rootTreeNode
        source.setQuotaInfo(Utils.getStorageStats(File(path)))
        return rootFile
    }

    override fun readFileTree(rootObject: Any?): TreeNode<SourceFile> {
        if (rootObject == null || rootObject !is File) {
            return rootTreeNode
        }
        val listFile = rootObject.listFiles()
        var dirSize = 0L
        if (listFile != null && listFile.isNotEmpty()) {
            for (file in listFile) {
                val sourceFile: SourceFile = LocalFile(file, source.sourceId)
                if (!file.isDirectory) {
                    dirSize += file.length()
                    currentNode.addChild(sourceFile)
                    continue
                }

                currentNode.addChild(sourceFile)
                currentNode = currentNode.children[currentNode.children.size - 1]
                readFileTree(file)

                currentNode.parent!!.data.size += currentNode.data.size
                currentNode = currentNode.parent!!

            }
            currentNode.data.size += dirSize
        }
        return rootTreeNode
    }
}